pkg/reconciler/namespacecontroller/namespace_controller.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"kpt.dev/configsync/pkg/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NamespaceController is a controller that watches for Namespace events and
+// sets a flag to indicate whether the Reconciler thread needs to reconcile for
+// this namespace event.
+type NamespaceController struct {
+	client client.Client
+	state  *State
+}
+
+// New instantiates the namespace controller.
+func New(cl client.Client, state *State) *NamespaceController {
+	return &NamespaceController{
+		client: cl,
+		state:  state,
+	}
+}
+
+// Reconcile processes the Namespace request to set the flag to indicate whether
+// a new reconciliation needs to be executed by the Reconciler thread.
+func (nc *NamespaceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ns := &corev1.Namespace{}
+	if err := nc.client.Get(ctx, req.NamespacedName, ns); err != nil {
+		if apierrors.IsNotFound(err) {
+			if nc.state.isSelectedNamespace(req.Name) {
+				// A previously selected namespace is deleted, so set the namespace event.
+				nc.state.setNewNSEvent()
+			}
+			// Ignore the event of the non-selected namespaces
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, status.APIServerError(err, "failed to get Namespace "+req.Name)
+	}
+
+	if nc.state.selectedOrMatchSelector(ns) {
+		// The created/updated namespace is either previously selected, or its
+		// labels match at least one NamespaceSelector's labels.
+		nc.state.setNewNSEvent()
```

Comment:
It looks like we only check that the labels match the selectors and don't check for any kind of diff we might care about. So ANY change to a matching namespace will trigger sync. We might want to avoid unnecessary syncs by checking if the set of matching namespaces has changed.

The deletion detection has the same problem, but we can probably reduce those with the same method.

These are the use cases, afaik:
1. if it used to be selected and is no longer selected, we need to sync
2. if it was not selected and now is selected, we need to sync
3. is it was selected and is now deleted, we need to sync.

So I think we need to cache the set of namespace that were selected as of the last successful sync. That way we can check if any change to that list has been made with the current state and only then trigger a new sync.

Reply:
The deletion detection seems to be okay. `apierrors.IsNotFound(err)` and `nc.state.isSelectedNamespace(req.Name)` matches use case 3.
The `selectedNamespaces` caches the set of namespaces that were selected, but may not always get to a successful sync.
We can change the value from an empty struct into the label selector, and detects whether it still matches or not.

Reply:
updated `selectedOrMatchSelector` to `matchChanged`

Reply:
It is intended to be updated when parsing the source, and the namespace controller is running.